vita3k/modules/SceKernelModulemgr/SceModulemgr.cpp

```diff
@@ -16,18 +16,111 @@
 // 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 
 #include "SceModulemgr.h"
+#include <host/load_self.h>
+#include <io/functions.h>
 #include <kernel/state.h>
+#include <kernel/thread/thread_functions.h>
+#include <modules/module_parent.h>
+#include <util/lock_and_find.h>
+
+static bool load_module(SceUID &mod_id, Ptr<const void> &entry_point, SceKernelModuleInfoPtr &module, HostState &host, const char *export_name, const char *path, int &error_val) {
+    const auto &loaded_modules = host.kernel.loaded_modules;
+
+    auto module_iter = std::find_if(loaded_modules.begin(), loaded_modules.end(), [path](const auto &p) {
+        return std::string(p.second->path) == path;
+    });
+
+    if (module_iter == loaded_modules.end()) {
+        // module is not loaded, load it here
+
+        const auto file = open_file(host.io, path, SCE_O_RDONLY, host.pref_path, export_name);
+        if (file < 0) {
+            error_val = RET_ERROR(file);
+            return false;
+        }
+        const auto size = seek_file(file, 0, SCE_SEEK_END, host.io, export_name);
+        if (size < 0) {
+            error_val = RET_ERROR(SCE_ERROR_ERRNO_EINVAL);
+            return false;
+        }
+
+        if (seek_file(file, 0, SCE_SEEK_SET, host.io, export_name) < 0) {
+            error_val = RET_ERROR(static_cast<int>(size));
+            return false;
+        }
+
+        auto *data = new char[static_cast<int>(size) + 1]; // null-terminated char array
```

Comment:
It would be nice if this was `std::vector<char>` or something, RAII would cleanup everything nicely.